### PR TITLE
Updating the testing package to v6

### PR DESCRIPTION
**Pull request by mat-mcloughlin**
Connects to https://github.com/Particular/NServiceBus.Testing/issues/29

### CODE REVIEW

File: src/NServiceBus.Testing.Tests/Issue508.cs [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
 {
     using System;
+    using System.Threading.Tasks;
     using MyMessages;
     using NUnit.Framework;
-    using Saga;
 
     [TestFixture]
-    public class Issue508 : BaseTests
+    public class Issue508
```
Comment (timbussmann) [inline src/NServiceBus.Testing.Tests/Issue508.cs]: I think we should get rid of these IssueXYZ tests. They are not obvious and if these tests add value, they should be with the rest of the tests.

File: src/NServiceBus.Testing/ExpectedInvocations/ExpectedInvocation.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    using System.Collections.Generic;
+    using System.Linq;
+
+    abstract class ExpectedInvocation
+    {
+        internal abstract void Validate(TestableMessageHandlerContext context);
+
+        internal void Fail<TMessage>(IList<TMessage> invokedMessages)
+        {
+            throw new Exception($"{Filter(GetType())} not fulfilled.\nCalls made:\n{string.Join("\n", invokedMessages.Select(i => Filter(i.GetType())))}");
```
Comment (timbussmann) [inline src/NServiceBus.Testing/ExpectedInvocations/ExpectedInvocation.cs]: does this error message still make sense in the v6 context?

File: src/NServiceBus.Testing.Tests/Issue508.cs (hunk 1 above)
Comment (mat-mcloughlin) [inline src/NServiceBus.Testing.Tests/Issue508.cs]: Yeah I agree. I'll move them in

**Comment by mat-mcloughlin** [comment]
@hmemcpy I have merged @timbussmann PR in where he handles the inverse expectations differently. Think I prefer it but need to decide and be consistent on a method

**Comment by mat-mcloughlin** [comment]
Good news. I've got all the tests passing in v6 now but there are still a couple of things I need to do

- Go through and do some code tidying
- Check to see if there are any additional tests we can add on top of the ones that were already there
- @timbussmann  we need to have a discussion about if we should be using obsolete methods as you removed the methods for `SendToSites` rather than obsoleting them so I want to be consistent

File: src/NServiceBus.Testing/Handler.cs [hunk 3]
```diff
@@ -105,16 +106,16 @@ public Handler<T> ExpectNotSendLocal<TMessage>(Func<TMessage, bool> check)
         /// </summary>
         public Handler<T> ExpectReturn<TEnum>(Func<TEnum, bool> check)
```
Comment (hmemcpy) [inline src/NServiceBus.Testing/Handler.cs]: This is the only remaining case without a test. I can't find previous usage. What does this test for?

Comment (timbussmann) [inline src/NServiceBus.Testing/Handler.cs]: This is from the callbacks API. There was a `Return` and a `Reply` which now is merged in `Reply`

File: src/NServiceBus.Testing.Tests/SagaTests.cs [hunk 4]
```diff
@@ -25,9 +25,8 @@ public void MySaga()
         public void MySagaWithActions()
         {
             Test.Saga<MySaga>()
-                .ExpectTimeoutToBeSetIn<StartsSaga>(
-                    (state, span) => Assert.That(() => span, Is.EqualTo(TimeSpan.FromDays(7))))
-                .When(s => s.Handle(new StartsSaga()));
+                .ExpectTimeoutToBeSetIn<StartsSaga>((state, span) => Assert.That(() => span, Is.EqualTo(TimeSpan.FromDays(7))))
```
Comment (timbussmann) [inline src/NServiceBus.Testing.Tests/SagaTests.cs]: seems wrong to use Assert in a check function?

File: src/NServiceBus.Testing.Tests/SagaTests.cs [hunk 5]
```diff
@@ -200,44 +333,47 @@ public class MyReply
...
+            await context.Send<Command>(s => { });
+            await context.ForwardCurrentMessageTo("forwardingDestination");
+            await RequestTimeout(context, TimeSpan.FromDays(7), message);
         }
 
-        public void Timeout(StartsSaga state)
+        public Task Timeout(StartsSaga state, IMessageHandlerContext context)
         {
-            Bus.Publish<Event>();
+            context.Publish<Event>();
```
Comment (timbussmann) [inline src/NServiceBus.Testing.Tests/SagaTests.cs]: missing await

File: src/NServiceBus.Testing.Tests/SagaTests.cs [hunk 6]
```diff
@@ -200,44 +333,47 @@ public class MyReply
...
             MarkAsComplete();
+
+            return Task.FromResult(0);
```
Comment (timbussmann) [inline src/NServiceBus.Testing.Tests/SagaTests.cs]: not needed

File: src/NServiceBus.Testing.Tests/SagaTests.cs [hunk 7]
```diff
@@ -200,44 +333,47 @@ public class MyReply
...
+    public class MySagaWithInterface : NServiceBus.Saga<MySagaWithInterface.MySagaDataWithInterface>,
         IAmStartedByMessages<StartsSagaWithInterface>
     {
-        public class MySagaDataWithInterface : ContainSagaData
+        public Task Handle(StartsSagaWithInterface message, IMessageHandlerContext context)
         {
-            
+            if (message.Foo == "Hello")
+            {
+                context.Send<Command>(m => { });
```
Comment (timbussmann) [inline src/NServiceBus.Testing.Tests/SagaTests.cs]: return task

File: src/NServiceBus.Testing.Tests/SagaTests.cs [hunk 8]
```diff
@@ -349,11 +492,52 @@ public class ProcessOrder : IMessage
...
+                           IHandleTimeouts<TheTimeout>
+    {
+        public Task Handle(TheMessage message, IMessageHandlerContext context)
+        {
+            return RequestTimeout<TheTimeout>(context, message.TimeoutAt);
+        }
+
+        public Task Timeout(TheTimeout state, IMessageHandlerContext context)
+        {
+            context.Send(new TheMessageSentAtTimeout());
```
Comment (timbussmann) [inline src/NServiceBus.Testing.Tests/SagaTests.cs]: missing await

File: src/NServiceBus.Testing.Tests/SagaTests.cs [hunk 9]
```diff
@@ -349,11 +492,52 @@ public class ProcessOrder : IMessage
...
+            MarkAsComplete();
+
+            return Task.FromResult(0);
```
Comment (timbussmann) [inline src/NServiceBus.Testing.Tests/SagaTests.cs]: remove

File: src/NServiceBus.Testing/ExpectedInvocations/ExpectHandleCurrentMessageLater.cs [hunk 10]
```diff
@@ -0,0 +1,13 @@
+﻿namespace NServiceBus.Testing.ExpectedInvocations
+{
+    class ExpectHandleCurrentMessageLater : ExpectInvocation
+    {
+        public override void Validate(TestableMessageHandlerContext context)
+        {
+            if (!context.HandleCurrentMessageLaterWasCalled)
+            {
+                Fail(context.SentMessages);
```
Comment (timbussmann) [inline src/NServiceBus.Testing/ExpectedInvocations/ExpectHandleCurrentMessageLater.cs]: Feels like this should be an entirely different text here since it's not related to outgoing messages at all?

File: src/NServiceBus.Testing/ExpectedInvocations/ExpectNotForwardCurrentMessageTo.cs [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+        public ExpectNotForwardCurrentMessageTo(Func<string, bool> check)
+        {
+            this.check = check;
+        }
+
+        public override void Validate(TestableMessageHandlerContext context)
+        {
+            if (context.ForwardedMessages.Any(m => check(m)))
+            {
+                Fail(Enumerable.Empty<object>());
```
Comment (timbussmann) [inline src/NServiceBus.Testing/ExpectedInvocations/ExpectNotForwardCurrentMessageTo.cs]: shouldn't the forwarded messages be passed to fail?

File: src/NServiceBus.Testing/ExpectedInvocations/ExpectReplyToOriginator.cs [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        public ExpectReplyToOriginator(Func<TMessage, bool> check = null)
+        {
+            this.check = check ?? (m => true);
+        }
+
+        public override void Validate(TestableMessageHandlerContext context)
+        {
+            var repliedMessages = context.RepliedMessages
+                .Where(i => i.Message.GetType() == typeof(TMessage))
```
Comment (timbussmann) [inline src/NServiceBus.Testing/ExpectedInvocations/ExpectReplyToOriginator.cs]: does this check work with generated proxy classes (e.g. when using Reply<IInterfaceType>(..)?) or should that check assignability?

Comment (hmemcpy) [inline src/NServiceBus.Testing/ExpectedInvocations/ExpectReplyToOriginator.cs]: I think this should be changed to the `Containing<TMessage>` extension method...

File: src/NServiceBus.Testing/Handler.cs [hunk 13]
```diff
@@ -4,24 +4,25 @@
...
     /// <summary>
     /// Message handler unit testing framework.
     /// </summary>
     public class Handler<T>
     {
-        private readonly StubBus bus;
         private readonly T handler;
-        private IDictionary<string, string> incomingHeaders = new Dictionary<string, string>();
-        private IList<IExpectedInvocation> expectedInvocations = new List<IExpectedInvocation>();
+        IMessageCreator messageCreator = new MessageMapper();
```
Comment (timbussmann) [inline src/NServiceBus.Testing/Handler.cs]: can we inline that into `TestableMessageHandlerContext`

File: src/NServiceBus.Testing/Handler.cs [hunk 14]
```diff
@@ -105,16 +106,16 @@ public Handler<T> ExpectNotSendLocal<TMessage>(Func<TMessage, bool> check)
...
-            expectedInvocations.Add(new ExpectedReturnInvocation<TEnum> { Check = check });
+            //testableMessageHandlerContext.ExpectedInvocations.Add(new ExpectedReturnInvocation<TEnum> { Check = check });
             return this;
         }
 
         /// <summary>
         /// Check that the handler sends the given message type to the appropriate destination.
         /// </summary>
-        public Handler<T> ExpectSendToDestination<TMessage>(Func<TMessage, Address, bool> check)
+        public Handler<T> ExpectSendToDestination<TMessage>(Func<TMessage, string, bool> check)
```
Comment (timbussmann) [inline src/NServiceBus.Testing/Handler.cs]: obsolete/remove

File: src/NServiceBus.Testing/Handler.cs [hunk 15]
```diff
@@ -241,38 +242,43 @@ public void OnMessage<TMessage>(TMessage initializedMessage)
...
+            var h = GetMessageHandler<TMessage>(handler.GetType(), message.GetType());
+            h.Invoke(handler, new object[] { message, testableMessageHandlerContext });
+
+            testableMessageHandlerContext.Validate();
+        }
+
+        /// <summary>
+        /// Check that the handler sends a message of the given type to sites.
+        /// </summary>
+        [ObsoleteEx(
```
Comment (timbussmann) [inline src/NServiceBus.Testing/Handler.cs]: should we add some sort of message to the obsoleted expectations?

File: src/NServiceBus.Testing/Saga.cs [hunk 16]
```diff
@@ -35,377 +31,362 @@ internal Saga(T saga, StubBus bus)
             saga.Entity.OriginalMessageId = Guid.NewGuid().ToString();
             saga.Entity.Originator = "client";
         }
+
         /// <summary>
-        /// Provides a way to set external dependencies on the saga under test.
+        ///     Set the address of the client that caused the saga to be started.
```
Comment (timbussmann) [inline src/NServiceBus.Testing/Saga.cs]: fix weird spacings

File: src/NServiceBus.Testing/Saga.cs [hunk 17]
```diff
@@ -35,377 +31,362 @@ internal Saga(T saga, StubBus bus)
...
             return this;
         }
 
         /// <summary>
-        /// Get the headers set by the saga when it sends a message.
+        ///     Check that the saga sends a message of the given type complying with user-supplied assertions.
         /// </summary>
-        public IDictionary<string, string> OutgoingHeaders
+        /// <param name="check">An action containing assertions on the message.</param>
+        public Saga<T> ExpectSend<TMessage>(Action<TMessage> check)
```
Comment (timbussmann) [inline src/NServiceBus.Testing/Saga.cs]: what is that `Action<TMessage> check` used for?

File: src/NServiceBus.Testing/Test.cs [hunk 18]
```diff
@@ -1,118 +1,28 @@
...
-        /// </summary>
-        public static IBus Bus
-        {
-            get { return bus; }
-        }
-
-        /// <summary>
-        ///     Initializes the testing infrastructure.
-        /// </summary>
-        public static void Initialize(Action<BusConfiguration> customisations = null)
```
Comment (timbussmann) [inline src/NServiceBus.Testing/Test.cs]: :dancer: :tada:

File: src/NServiceBus.Testing/TestableMessageHandlerContext.cs [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+                Clear();
+            }
+        }
+
+        public void Clear()
+        {
+            ExpectedInvocations.Clear();
+            //RepliedMessages.Clear();
+            //SentMessages.Clear();
+            //PublishedMessages.Clear();
```
Comment (timbussmann) [inline src/NServiceBus.Testing/TestableMessageHandlerContext.cs]: why are they commented out?

File: src/NServiceBus.Testing/ExpectedInvocations/ExpectHandleCurrentMessageLater.cs (hunk 10 above)
Comment (mat-mcloughlin) [inline src/NServiceBus.Testing/ExpectedInvocations/ExpectHandleCurrentMessageLater.cs]: I'm not sure I follow. Are you talking about the exception message that is thrown?

File: src/NServiceBus.Testing/Handler.cs (hunk 13 above)
Comment (mat-mcloughlin) [inline src/NServiceBus.Testing/Handler.cs]: We could but I believe @hmemcpy explicitly did it this way to avoid creating two instances?

Comment (hmemcpy) [inline src/NServiceBus.Testing/Handler.cs]: @mat-mcloughlin @timbussmann I wanted to keep *some* inversion here, so that the `TestableMessageHandlerContext` is highly composable in testing. But, I don't mind inlining it, if that helps.

File: src/NServiceBus.Testing/TestableMessageHandlerContext.cs (hunk 19 above)
Comment (mat-mcloughlin) [inline src/NServiceBus.Testing/TestableMessageHandlerContext.cs]: My bad, should never have been there in the first place. Its from me trying to figure stuff out

File: src/NServiceBus.Testing/Test.cs [hunk 20]
```diff
@@ -140,75 +50,34 @@ public static Saga<T> Saga<T>(Guid sagaId) where T : Saga, new()
...
-                             where typeof(IHandleMessages<>).MakeGenericType(args[0]).IsAssignableFrom(i)
-                             select i).Any();
-
-            if (!isHandler)
-            {
-                throw new ArgumentException("The handler object created does not implement IHandleMessages<T>.", "handlerCreationCallback");
-            }
-
-            return new Handler<T>(handler, bus);
+            throw new NotImplementedException();
```
Comment (mat-mcloughlin) [inline src/NServiceBus.Testing/Test.cs]: @timbussmann @hmemcpy It feels like this method is no longer required as we don't need to give them a bus to inject into the constructor. Are you happy obsolete it?

Comment (timbussmann) [inline src/NServiceBus.Testing/Test.cs]: definitely obsolete it :+1:

File: src/NServiceBus.Testing/Handler.cs (hunk 3 above)
Comment (hmemcpy) [inline src/NServiceBus.Testing/Handler.cs]: So obsolete this one as well? @mat-mcloughlin @timbussmann

File: src/NServiceBus.Testing/ExpectedInvocations/ExpectHandleCurrentMessageLater.cs (hunk 10 above)
Comment (hmemcpy) [inline src/NServiceBus.Testing/ExpectedInvocations/ExpectHandleCurrentMessageLater.cs]: It seems that Fail should have an overload that takes a message to display? @mat-mcloughlin @timbussmann

Comment (timbussmann) [inline src/NServiceBus.Testing/ExpectedInvocations/ExpectHandleCurrentMessageLater.cs]: SentMessages isn't related to the fact that `HandleCurrentMessageLater` called or not.

**Comment by timbussmann** [comment]
@Particular/nservicebus-maintainers can you have a look too?

**Comment by timbussmann** [comment]
Should we provide a `Saga<T>(..)` overload which accepts a saga data instance?

File: src/NServiceBus.Testing/TestableMessageHandlerContext.cs [hunk 21]
```diff
@@ -0,0 +1,164 @@
...
+            HandleCurrentMessageLaterWasCalled = true;
+            return Task.FromResult(0);
+        }
+
+        public void DoNotContinueDispatchingCurrentMessageToHandlers()
+        {
+            HandlerInvocationAborted = true;
+        }
+
+        public Task Subscribe(Type eventType, SubscribeOptions options)
```
Comment (timbussmann) [inline src/NServiceBus.Testing/TestableMessageHandlerContext.cs]: I think we can remove `Subscribe` and `Unsubscribe` since they are no longer supported in the `IPipelineContext`

File: src/NServiceBus.Testing/Test.cs [hunk 22]
```diff
@@ -216,19 +75,27 @@ public static TMessage CreateInstance<TMessage>()
...
-        static StubBus bus;
+        /// <summary>
+        /// Specify a test for a message handler specifying a callback to create
+        /// the handler and getting an instance of the bus passed in.
+        /// Useful for handlers based on constructor injection.
+        /// </summary>
+        [ObsoleteEx(
+            RemoveInVersion = "7",
+            TreatAsErrorFromVersion = "6")]
+        public static Handler<T> Handler<T>(Func<IMessageSession, T> handlerCreationCallback)
```
Comment (timbussmann) [inline src/NServiceBus.Testing/Test.cs]: the obsoleted signature should be `Handler<T> Handler<T>(Func<IBus, T> handlerCreationCallback)` not `IMessageSession`

File: src/NServiceBus.Testing/Handler.cs (hunk 3 above)
Comment (timbussmann) [inline src/NServiceBus.Testing/Handler.cs]: yes

**Comment by mat-mcloughlin** [comment]
I'm not saying we shouldn't but there wasn't one before. Perhaps this is something we can consider when looking at improvements?

**Comment by timbussmann** [comment]
I can live with that. Put it on my list.

File: src/NServiceBus.Testing/Test.cs (hunk 22 above)
Comment (danielmarbach) [inline src/NServiceBus.Testing/Test.cs]: :8ball: https://github.com/Particular/NServiceBus.Testing/pull/43#discussion-diff-54435776L227

**Comment by danielmarbach** [comment]
Can we squash this?

**Comment by mat-mcloughlin** [comment]
Yep, just got one small change to add and I will do it in the morning

**Comment by mat-mcloughlin** [comment]
@danielmarbach @Particular/nservicebus-maintainers squashed and ready for review

File: src/NServiceBus.Testing.Tests/HandlerTests.cs [hunk 23]
```diff
@@ -461,110 +550,90 @@ public class Incoming : IMessage
...
             public TimeSpan Defer { get; set; }
 
-            public void Handle(TestMessage message)
+            public async Task Handle(TestMessage message, IMessageHandlerContext context)
             {
-                Bus.Defer(Defer, message);
+                var sendOptions = new SendOptions();
+                sendOptions.DelayDeliveryWith(Defer);
+
+                await context.Send(message, sendOptions);
```
Comment (danielmarbach) [inline src/NServiceBus.Testing.Tests/HandlerTests.cs]: Down below you have a few handler which only have on async operation in the handler and there you return the context task. Would it make sense to do the same here?

File: src/NServiceBus.Testing.Tests/HandlerTests.cs [hunk 24]
```diff
@@ -461,110 +550,90 @@ public class Incoming : IMessage
...
             public DateTime Defer { get; set; }
 
-            public void Handle(TestMessage message)
+            public async Task Handle(TestMessage message, IMessageHandlerContext context)
             {
-                Bus.Defer(Defer, message);
+                var sendOptions = new SendOptions();
+                sendOptions.DoNotDeliverBefore(Defer);
+
+                await context.Send(message, sendOptions);
```
Comment (danielmarbach) [inline src/NServiceBus.Testing.Tests/HandlerTests.cs]: Down below you have a few handler which only have on async operation in the handler and there you return the context task. Would it make sense to do the same here?

File: src/NServiceBus.Testing.Tests/SagaTests.cs [hunk 25]
```diff
@@ -168,31 +192,140 @@ public void ShouldPassExpectNotForwardCurrentMessageToIfMessageForwardedToUnexpe
...
+        }
 
-    public class SagaThatDoesAReply : Saga<SagaThatDoesAReply.SagaThatDoesAReplyData>,
+        public class Message
+        {
+            public string Property => "Property";
+        }
+    }
+
+    public class HandleInFutureSaga : NServiceBus.Saga<HandleInFutureSaga.HandleInFutureSagaData>,
```
Comment (danielmarbach) [inline src/NServiceBus.Testing.Tests/SagaTests.cs]: Is the `NServiceBus.Saga` (full namespace) really necessary?

File: src/NServiceBus.Testing.Tests/SagaTests.cs [hunk 26]
```diff
@@ -168,31 +192,140 @@ public void ShouldPassExpectNotForwardCurrentMessageToIfMessageForwardedToUnexpe
...
         }
 
-        public void Handle(MyRequest myRequest)
+        public class HandleInFutureSagaData : ContainSagaData
         {
-            Bus.Reply(new MyReply());
+        }
+    }
+
+    public class SagaThatDoesAReply : NServiceBus.Saga<SagaThatDoesAReply.SagaThatDoesAReplyData>,
```
Comment (danielmarbach) [inline src/NServiceBus.Testing.Tests/SagaTests.cs]: Is the `NServiceBus.Saga` (full namespace) really necessary?

Comment (danielmarbach) [inline src/NServiceBus.Testing.Tests/SagaTests.cs]: We have a few other fully qualified namespaces.

File: src/NServiceBus.Testing.sln.DotSettings [hunk 27]
```diff
@@ -32,6 +32,8 @@
 	<s:String x:Key="/Default/CodeInspection/Highlighting/InspectionSeverities/=CSharpWarnings_003A_003ACS1584/@EntryIndexedValue">ERROR</s:String>
 	<s:String x:Key="/Default/CodeInspection/Highlighting/InspectionSeverities/=CSharpWarnings_003A_003ACS1587/@EntryIndexedValue">ERROR</s:String>
 	<s:String x:Key="/Default/CodeInspection/Highlighting/InspectionSeverities/=CSharpWarnings_003A_003ACS1589/@EntryIndexedValue">ERROR</s:String>
+	<s:String x:Key="/Default/CodeInspection/Highlighting/InspectionSeverities/=CSharpWarnings_003A_003ACS1591/@EntryIndexedValue"></s:String>
```
Comment (danielmarbach) [inline src/NServiceBus.Testing.sln.DotSettings]: Is this coming from RepoStandards?

File: src/NServiceBus.Testing/ExpectedInvocations/ExpectInvocation.cs [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            throw new Exception($"{Filter(GetType())} not fulfilled.\nCalls made:\n{string.Join("\n", invokedMessages.Select(i => Filter(i.GetType())))}");
+        }
+
+        protected void Fail(string message)
+        {
+            throw new Exception(message);
+        }
+
+        string Filter(Type type)
```
Comment (danielmarbach) [inline src/NServiceBus.Testing/ExpectedInvocations/ExpectInvocation.cs]: Should we give this method a better name expressing what it does?

File: src/NServiceBus.Testing/ExpectedInvocations/ExpectNotDelayDeliveryWith.cs [hunk 29]
```diff
@@ -0,0 +1,29 @@
...
+        {
+            var sentMessages = context.SentMessages
+                .Containing<TMessage>()
+                .Where(s => s.Options.GetDeliveryDelay().HasValue)
+                .ToList();
+
+            if (sentMessages.Any(s => check(s.Message, s.Options.GetDeliveryDelay().Value)))
+            {
+                Fail(sentMessages.Select(i => i.Message).ToList());
+
```
Comment (danielmarbach) [inline src/NServiceBus.Testing/ExpectedInvocations/ExpectNotDelayDeliveryWith.cs]: Unnecessary newline

File: src/NServiceBus.Testing/ExpectedInvocations/ExpectSendToDestination.cs [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+
+        public override void Validate(TestableMessageHandlerContext context)
+        {
+            var sentMessages = context.SentMessages
+                .Where(i => i.Message.GetType() == typeof(TMessage))
+                .Where(i => string.IsNullOrWhiteSpace(i.Options.GetCorrelationId()) &&
+                    !string.IsNullOrWhiteSpace(i.Options.GetDestination()))
+                .ToList();
+
+            if (!sentMessages.Any(i => check((TMessage) i.Message, i.Options.GetDestination())))
```
Comment (danielmarbach) [inline src/NServiceBus.Testing/ExpectedInvocations/ExpectSendToDestination.cs]: Do we really need two casts? One here and one with LINQ `Cast<>`

File: src/NServiceBus.Testing/Handler.cs [hunk 31]
```diff
@@ -105,16 +108,7 @@ public Handler<T> ExpectNotSendLocal<TMessage>(Func<TMessage, bool> check)
...
-            return this;
-        }
-
-        /// <summary>
-        /// Check that the handler sends the given message type to the appropriate destination.
-        /// </summary>
-        public Handler<T> ExpectSendToDestination<TMessage>(Func<TMessage, Address, bool> check)
-        {
-            expectedInvocations.Add(new ExpectedSendToDestinationInvocation<TMessage> { Check = check });
+            //testableMessageHandlerContext.ExpectedInvocations.Add(new ExpectedReturnInvocation<TEnum> { Check = check });
```
Comment (danielmarbach) [inline src/NServiceBus.Testing/Handler.cs]: Commented out code?

File: src/NServiceBus.Testing/Handler.cs [hunk 32]
```diff
@@ -241,41 +235,68 @@ public void OnMessage<TMessage>(TMessage initializedMessage)
         /// </summary>
         public void OnMessage<TMessage>(TMessage message, string messageId)
         {
-            var context = new MessageContext { Id = messageId, ReturnAddress = "client", Headers = incomingHeaders };
-            bus.CurrentMessageContext = context;
+            testableMessageHandlerContext.MessageId = messageId;
 
-            foreach (var keyValuePair in incomingHeaders)
+            var h = GetMessageHandler<TMessage>(handler.GetType(), message.GetType());
+            h.Invoke(handler, new object[]
```
Comment (danielmarbach) [inline src/NServiceBus.Testing/Handler.cs]: Something is off. Let me check in the code directly.

File: src/NServiceBus.Testing.Tests/SagaTests.cs (hunk 26 above)
Comment (timbussmann) [inline src/NServiceBus.Testing.Tests/SagaTests.cs]: I think yes since it collides with the `NServiceBus.Testing.Saga` class?

File: src/NServiceBus.Testing.sln.DotSettings (hunk 27 above)
Comment (timbussmann) [inline src/NServiceBus.Testing.sln.DotSettings]: nope don't think so. I think I will update them anyway on develop after we merged this

File: src/NServiceBus.Testing/Handler.cs (hunk 31 above)
Comment (timbussmann) [inline src/NServiceBus.Testing/Handler.cs]: this should be obsoleted since SendToDestination belongs to the Gateway package.

File: src/NServiceBus.Testing/TestableMessageHandlerContext.cs [hunk 33]
```diff
@@ -0,0 +1,162 @@
...
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.ObjectModel;
+    using System.Threading.Tasks;
+    using NServiceBus.Extensibility;
+    using NServiceBus.Persistence;
+    using NServiceBus.Testing.ExpectedInvocations;
+
+    internal class TestableMessageHandlerContext : IMessageHandlerContext
```
Comment (hmemcpy) [inline src/NServiceBus.Testing/TestableMessageHandlerContext.cs]: Should this class be `internal`? Are we planning to showcase it as the main test double for the handler context?

Comment (timbussmann) [inline src/NServiceBus.Testing/TestableMessageHandlerContext.cs]: I think we'll make this one public as a separate PR since this class seems to need some more work in order to achieve that.

**Comment by timbussmann** [comment]
merged2
